### PR TITLE
USB: Unifying legacy and composite (including cherry-picked)

### PR DESCRIPTION
**Pull request by finikorg**
Cherry-picked patches from PR #11041 (which was closed). Rebased against latest master.

@aurel32 please push any missing patches to this branch if needed.

### CODE REVIEW

**Review by jfischer-no — CHANGES_REQUESTED** [review]
LGTM, except parts regarding default endpoint addresses (AUTO_EP_*)

**Review by dbkinder — APPROVED** [review]
Doc changes LGTM.

**Comment by jfischer-phytec-iot** [comment]
@finikorg Is it still WIP?

**Review by jfischer-no — CHANGES_REQUESTED** [review]
LGTM, just few comments

File: subsys/usb/class/netusb/netusb.c [hunk 1]
```diff
@@ -137,34 +137,6 @@ static void netusb_init(struct net_if *iface)
 
 	net_if_down(iface);
 
-#ifndef CONFIG_USB_COMPOSITE_DEVICE
```
[the remark this answers is not in the document]
Comment (finikorg) [reply subsys/usb/class/netusb/netusb.c]: done

File: tests/subsys/usb/desc_sections/src/desc_sections.c [hunk 2]
```diff
@@ -221,7 +221,7 @@ static void test_desc_sections(void)
 			(int)__usb_data_end - (int)__usb_data_start,
 			"USB Configuratio structures section");
 
-	head = (struct usb_desc_header *)usb_get_device_descriptor();
+	head = (struct usb_desc_header *)__usb_descriptor_start;
```
[the remark this answers is not in the document]
Comment (finikorg) [reply tests/subsys/usb/desc_sections/src/desc_sections.c]: done

**Comment by finikorg** [comment]
Rebased and fixed conflicts with master.

**Comment by jfischer-phytec-iot** [comment]
Tested it for a while:
hci_usb :x: (on master :x:)
cdc_acm :white_check_mark: 
cdc_acm_composite :white_check_mark: 
dfu :white_check_mark: 
dfu + composite mode :x: (class_handler (and custom_handler) is hardcoded to primary set of descriptors, dfu has two set of descriptors and does not work in composite configuration)
hid-mouse :white_check_mark: 
hid-cdc :white_check_mark: 
mass :white_check_mark: 
testusb :white_check_mark:

**Comment by masz-nordic** [comment]
Tested with `nrf52840_pca10056`:

|       Sample      |       Result       |
|:-----------------:|:------------------:|
|      cdc_acm      | :heavy_check_mark: |
| cdc_acm_composite | :heavy_check_mark: |
|        hid        | :heavy_check_mark: |
|     hid-mouse     | :heavy_check_mark: |
|      hid-cdc      | :heavy_check_mark: |
|        mass       | :heavy_check_mark: |

Although all of them exhibited reconnection issues, as expected.

**Review by masz-nordic — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Pull request merged**